frontend: expandable sidebar

adilb99 opened the pull request:
video of how it looks:
https://vimeo.com/1036801461/fe4c0a66c2?share=copy

A very hacky solution to having an expandable sidebar, but it works well on local so far. To do:

- [x] maybe a more robust check for whether it's a project page or a root page @ frontend/components/ui/header.tsx:26-28 (or use a different header component in the project page to have the SidebarTrigger properly in the header without affecting the root layout page).
- [x] less hacky styling cause currently the sidebar.tsx uses some of the styles from the old navbar directly instead of a reusable way.
- [x] resolve conflicts?
- [ ] apply ESLint rules consistently



<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Introduces an expandable sidebar with mobile support, using new `Sidebar` components and updated styling across the frontend.
> 
>   - **Behavior**:
>     - Introduces expandable sidebar using `SidebarProvider` and `Sidebar` components in `layout.tsx` and `project/[projectId]/layout.tsx`.
>     - Sidebar can be toggled using `SidebarTrigger` in `header.tsx`.
>     - Mobile support added with `useIsMobile` hook in `use-mobile.tsx`.
>   - **Components**:
>     - New `Sidebar` components in `sidebar.tsx` for layout and interaction.
>     - `ProjectNavbar` updated to use `Sidebar` components in `project-navbar.tsx`.
>     - `Sheet` components in `sheet.tsx` used for mobile sidebar.
>   - **Styling**:
>     - New CSS variables for sidebar styling in `globals.css`.
>     - Tailwind configuration updated in `tailwind.config.js` and `tailwind.config.ts` to support new styles.
>   - **Misc**:
>     - Minor refactoring in `pipeline-version-context.tsx` and `project-context.tsx` for consistency.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=lmnr-ai%2Flmnr&utm_source=github&utm_medium=referral)<sup> for 59c642afe47e917022f0e36c7b0e646d286c9339. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/lmnr-ai/lmnr?pullRequest=273) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/lmnr-ai/lmnr?pullRequest=273) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: skull8888888<br/>:x: adilb99<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/lmnr-ai/lmnr?pullRequest=273) it.</sub>

dinmukhamedm commented on frontend/app/globals.css:
```diff
@@ -70,6 +70,14 @@
     --chart-3: 30 80% 55%;
     --chart-4: 280 65% 60%;
     --chart-5: 340 75% 55%;
+    --sidebar-background: 0 0% 98%;
+    --sidebar-foreground: 240 5.3% 26.1%;
+    --sidebar-primary: 240 5.9% 10%;
+    --sidebar-primary-foreground: 0 0% 98%;
+    --sidebar-accent: 240 4.8% 95.9%;
+    --sidebar-accent-foreground: 240 5.9% 10%;
+    --sidebar-border: 220 13% 91%;
+    --sidebar-ring: 217.2 91.2% 59.8%;
```
are all these additional colors required by shadcn's `<Sidebar>`. Otherwise, I don't think we need them all, and could just tailwind all the necessary colors on the sidebar

dinmukhamedm replied:
@adilb99 Ah I see, you're saying we've already had these colors defined somewhere? Yeah, let's do classNames that we already have

dinmukhamedm commented on frontend/assets/logo/laminar.svg:
we'll add the new logo anw I guess

skull8888888 replied:
we already have new logo, check out how we do it in projects header and on landing header

skull8888888 replied:
so we should remove the old one

dinmukhamedm commented on frontend/components/project/project-navbar.tsx:
```diff
@@ -37,52 +56,75 @@ export default function ProjectNavbar({ projectId }: ProjectNavBarProps) {
-      </div>
-    </div>
+    <Sidebar className="border-r" collapsible='icon'>
+      <SidebarHeader className="h-14 bg-background">
```
I'd say we center this: `mx-auto`? But again, let's wait for Robert's input

dinmukhamedm commented on frontend/components/project/project-navbar.tsx:
```diff
@@ -37,52 +56,75 @@ export default function ProjectNavbar({ projectId }: ProjectNavBarProps) {
+        </SidebarMenu>
+        <div className="flex-grow" />
+        <div className="p-2">
+          <AvatarMenu />
```
avatar in the corner of an expanded sidebar looks a bit too empty. Should we maybe add an email to the right of it?

skull8888888 replied:
name and email underneath.

dinmukhamedm commented on frontend/components/ui/sidebar.tsx:
```diff
@@ -0,0 +1,745 @@
+            }
+
+            // This sets the cookie to keep the sidebar state.
+            document.cookie = `${SIDEBAR_COOKIE_NAME}=${openState}; path=/; max-age=${SIDEBAR_COOKIE_MAX_AGE}`;
```
Good one, but this is auto-downloaded from shadcn

dinmukhamedm commented on frontend/tailwind.config.js:
ouch, what's up with indents here?

dinmukhamedm reviewed (commented):
LGTM overall, left a few nitpicks, and there is clearly some more work to do

skull8888888 commented on frontend/components/ui/header.tsx:
```diff
@@ -22,8 +23,11 @@ export default function Header({ path, children, className }: HeaderProps) {
       className={`font-medium flex items-center justify-between flex-none h-12 border-b w-full ${className}`}
     >
       <div className="flex items-center">
+        {projectId.length > 0 && (
```
just add a prop, something like showSidebarButton and default to true, on /projects set it to false.

skull8888888 commented on frontend/components/project/project-navbar.tsx:
```diff
@@ -3,31 +3,50 @@
 import {
```
sidebar is too large, lots of empty space on the right to the text, let's make it smaller, around to the size of the largest text

adilb99 commented:
Addressed the comments. Gist:

- Better props in the header and avatar components
- Added email to the right of the avatar picture when expanded (username omitted due to being same as email)
- Narrower sidebar when expanded, aligned with the email
- Using new logo on expanded, deleted the old logo file + better naming for imported logos
- Decided to keep the sidebar css classes due to most of them still being used by sidebar.tsx

Let me know if there are any more improvements to be made. Otherwise, should be good to go!

skull8888888 commented:
Looks good, few things.
- State of the sidebar should be saved. Seems like sidebar should already have this kind of functionality judging by the fact that it has some cookies stuff. If it doesn't let's just use localStorage
- padding on the avatar should be consistent with the rest of the sidebar, seems like sidebar has p-4 but avatar p-2?
<img width="223" alt="Screenshot 2024-12-07 at 12 37 16 AM" src="https://github.com/user-attachments/assets/d87e36bf-2640-4a14-bb90-4f41d57dc8bb">

- let's actually make max width of the sidebar to the width of the largest option, in this case it's playgrounds. Let's truncate email. Because aligning max size with email would cause sidebar to be very large if a user has a very long email.
- button should align with the rest of the content which has p-4. Align button so that it aligns with the rest of the content.
<img width="464" alt="Screenshot 2024-12-07 at 12 42 26 AM" src="https://github.com/user-attachments/assets/3b2103f0-37b5-495a-96b4-4e66722b8073">

In general look at inconsistencies in the alignment, they immediately strike eyes.

adilb99 commented:
Fixed the following:

- Smaller sidebar width closer aligned with Playgrounds menu item
- Fixed padding and alignment issues
- Sidebar state is saved in the cookies and properly used. (see the next comment for details)

adilb99 commented on frontend/components/ui/sidebar.tsx:
```diff
@@ -70,10 +70,27 @@ const SidebarProvider = React.forwardRef<
         const [_open, _setOpen] = React.useState(defaultOpen);
         const open = openProp ?? _open;
+
+        // Move cookie check to useEffect to wait for document to be populated
+        React.useEffect(() => {
+          const sidebarCookie = document.cookie
+            .split('; ')
+            .find(row => row.startsWith(SIDEBAR_COOKIE_NAME))
+            ?.split('=')[1];
+
+          if (sidebarCookie !== undefined) {
+            const savedState = sidebarCookie === 'true';
+            if (setOpenProp) {
+              setOpenProp(savedState);
+            } else {
+              _setOpen(savedState);
+            }
+          }
+        }, [setOpenProp]); // Only runs once on mount
```
This is needed to handle SSR properly and not use `document` var before it is populated. However, this produces a slight lag when applying sidebar state on reload. Same occurs with proper usage of localStorage.

Possible solution is to retrieve the cookies on the server-side in layout.tsx and passing it into the component as a prop, but then propagating it back server-side to properly save makes the code a little too complex.

Possibly not a problem cause local frontend is a little slow and won't be so visible in production environment, but let me know your thoughts. @skull8888888 @dinmukhamedm

dinmukhamedm reviewed (commented):
LGTM. Not explicitly approving as @skull8888888 has this checked out, so he'll have another look before we merge

skull8888888 commented on frontend/tailwind.config.js:
```diff
@@ -10,83 +10,92 @@ module.exports = {
+        'accordion-up': 'accordion-up 0.2s ease-out'
       },
       fontSize: {
-        'lg': ['1.125rem', { lineHeight: '1.6rem', letterSpacing: '-0.0125em' }],
```
bring 'lg' and 'xl' back and put them in comas for consistency

skull8888888 reviewed (changes requested):
last change, otherwise lgtm
